### PR TITLE
*: add metrics for unified read pool

### PR DESCRIPTION
**Pull request by sticnarf**
###  What have you changed?

1. Flush storage and coprocessor metrics by tick in unified read pool.
2. Use newer version of yatp with metrics about multilevel thread pool. Because https://github.com/tikv/yatp/pull/27 is not merged, a patch is used here.

With these metrics we can see how the multilevel thread pool is scheduling small and big requests:

image

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Manual test (add detailed scripts or steps below)

Metrics are now available.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: cmd/src/server.rs [hunk 1]
```diff
@@ -281,6 +276,17 @@ impl TiKVServer {
         .unwrap();
     }
 
+    fn init_yatp(&self) {
+        let yatp_enabled = self.config.readpool.unify_read_pool;
```
Comment (youjiali1995) [inline cmd/src/server.rs]: Why not
```rust
        if !self.config.readpool.unify_read_pool {
            return;
        }
```

File: src/coprocessor/metrics.rs [hunk 2]
```diff
@@ -61,3 +68,88 @@ lazy_static! {
     )
     .unwrap();
 }
+
+pub struct CopLocalMetrics {
```
Comment (youjiali1995) [inline src/coprocessor/metrics.rs]: Why not move it to `src/coprocessor/local_metrics.rs`?

Comment (youjiali1995) [reply src/coprocessor/metrics.rs]: Oh, I found `src/coprocessor/local_metrics.rs` is never used..

File: src/read_pool.rs [hunk 3]
```diff
@@ -112,13 +121,50 @@ impl<E: Engine> Runner for ReadPoolRunner<E> {
...
+        }
+
+        THREAD_LAST_TICK_TIME.with(|tls_last_tick| {
+            let now = Instant::now_coarse();
+            let last_tick = tls_last_tick.get();
+            if now.duration_since(last_tick) < TICK_INTERVAL {
+                return;
+            }
+            tls_last_tick.set(now);
+            crate::storage::metrics::tls_flush(&self.reporter);
```
Comment (youjiali1995) [inline src/read_pool.rs]: Should flush them when the read pool ends.

File: cmd/src/server.rs (hunk 1 above)
Comment (sticnarf) [reply cmd/src/server.rs]: I just want to leave a hint that if there are other options which will enable yatp in the future, the options should be added here.

File: src/read_pool.rs [hunk 4]
```diff
@@ -108,17 +117,60 @@ impl<E: Engine> Runner for ReadPoolRunner<E> {
...
+    pub fn new(engine: E, inner: FutureRunner, reporter: R) -> Self {
         ReadPoolRunner {
             engine: Some(engine),
+            reporter,
             inner,
         }
     }
+
+    // Only flush metrics by tick
+    fn maybe_flush_metrics(&self) {
```
Comment (andylokandy) [inline src/read_pool.rs]: Would `try_flush_metrics` be better?

Comment (sticnarf) [reply src/read_pool.rs]: `try_` usually returns a `Result` while here we just ignore too frequent flushes

Comment (sticnarf) [reply src/read_pool.rs]: fixed. PTAL again.

**Pull request merged**